src/aws_encryption_sdk/keyrings/aws_kms/__init__.py

```diff
@@ -0,0 +1,376 @@
+            _AwsKmsSingleCmkKeyring(
+                key_id=key_id, client_supplier=self._client_supplier, grant_tokens=self._grant_tokens
+            )
+            for key_id in self._child_key_ids
+        ]
+
+        self._is_discovery = generator_keyring is None and not child_keyrings
+
+        if self._is_discovery:
+            self._inner_keyring = _AwsKmsDiscoveryKeyring(
+                client_supplier=self._client_supplier, grant_tokens=self._grant_tokens
+            )
+        else:
+            self._inner_keyring = MultiKeyring(generator=generator_keyring, children=child_keyrings)
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        """Generate a data key using generator keyring
+        and encrypt it using any available wrapping key in any child keyring.
+
+        :param EncryptionMaterials encryption_materials: Encryption materials for keyring to modify.
+        :returns: Optionally modified encryption materials.
+        :rtype: EncryptionMaterials
+        :raises EncryptKeyError: if unable to encrypt data key.
+        """
+        return self._inner_keyring.on_encrypt(encryption_materials=encryption_materials)
+
+    def on_decrypt(self, decryption_materials, encrypted_data_keys):
+        # type: (DecryptionMaterials, Iterable[EncryptedDataKey]) -> DecryptionMaterials
+        """Attempt to decrypt the encrypted data keys.
+
+        :param DecryptionMaterials decryption_materials: Decryption materials for keyring to modify.
+        :param List[EncryptedDataKey] encrypted_data_keys: List of encrypted data keys.
+        :returns: Optionally modified decryption materials.
+        :rtype: DecryptionMaterials
+        """
+        return self._inner_keyring.on_decrypt(
+            decryption_materials=decryption_materials, encrypted_data_keys=encrypted_data_keys
+        )
+
+
+@attr.s
+class _AwsKmsSingleCmkKeyring(Keyring):
+    """AWS KMS keyring that only works with a single AWS KMS CMK.
+
+    This keyring should never be used directly.
+    It should only ever be used internally by :class:`KmsKeyring`.
+
+    .. versionadded:: 1.5.0
+
+    :param str key_id: CMK key ID
+    :param ClientSupplier client_supplier: Client supplier to use when asking for clients
+    :param List[str] grant_tokens: AWS KMS grant tokens to include in requests (optional)
+    """
+
+    _key_id = attr.ib(validator=instance_of(six.string_types))
+    _client_supplier = attr.ib(validator=instance_of(ClientSupplier))
+    _grant_tokens = attr.ib(
+        default=attr.Factory(tuple),
+        validator=(deep_iterable(member_validator=instance_of(six.string_types)), value_is_not_a_string),
+    )
+
+    def on_encrypt(self, encryption_materials):
+        # type: (EncryptionMaterials) -> EncryptionMaterials
+        trace_info = MasterKeyInfo(provider_id=_PROVIDER_ID, key_info=self._key_id)
+        try:
+            if encryption_materials.data_encryption_key is None:
```

Comment:
This makes the keyring rely on the `MultiKeyring` to enforce the generator relationship.
I think I'm ok with this,
but I at least want to bring it up.

Reply:
This means that if this `_AwsKmsSingleCmkKeyring` is used in a MultiKeyring
*not* in generator position,
and this keyring is used as a generator for a parent MultiKeyring,
This combined keyring will **not** "do the right thing".

Will talk more.
But I want this to get documented.

Reply:
That is addressed by the multi-keyring. That's one reason why I chose to use the multi-keyring as an internal construct; aside from discovery mode, the requirements are identical.

https://github.com/aws/aws-encryption-sdk-python/blob/71b795dccb280b2b25b92c359d80c85623f2bd27/src/aws_encryption_sdk/keyrings/multi.py#L63-L77

Reply:
Yup, thanks.